### PR TITLE
Integrate CaDiCaL as CDCL(T) SAT solver.

### PR DESCRIPTION
**Pull request by aniemetz**
This integrates CaDiCaL via the IPASIR-UP interface as CDCL(T) SAT solver.
The CDCL(T) sat solver is configured via `--sat-solver`, default is MiniSat. If
proof production is enabled but CaDiCaL is configured, we fall back to MiniSat.

### CODE REVIEW

File: src/options/prop_options.toml [hunk 1]
```diff
@@ -1,6 +1,19 @@
 id     = "PROP"
 name   = "SAT Layer"
 
+[[option]]
+  name       = "satSolver"
+  category   = "expert"
+  long       = "sat-solver=MODE"
+  type       = "SatSolverMode"
+  default    = "MINISAT"
```
Comment (barrettcw) [inline src/options/prop_options.toml:9]: Why all caps here?

**Review by barrettcw — APPROVED** [review]
Obviously, I'm not going to look through every line of code in the cadical solver.  The small changes elsewhere look fine to me, but clearly, with a change of this magnitude, extensive testing will be the key.

File: src/options/prop_options.toml (hunk 1 above)
Comment (aniemetz) [reply src/options/prop_options.toml:9]: This is the name of the corresponding enum value, e.g., `SatSolverMode::MINISAT`. That's why it has to be upper case.

Comment (barrettcw) [reply src/options/prop_options.toml:9]: Got it.

File: src/prop/theory_preregistrar.h [hunk 2]
```diff
@@ -61,9 +61,8 @@ class TheoryPreregistrar : protected EnvObj
    */
   void notifySatLiteral(TNode n);
   /**
-   * Callback to notify that the SAT solver backtracked by the given number
-   * of levels.
-   * @param nlevels The number of levels the SAT solver backtracked.
+   * Callback to notify that the SAT solver backtracked to the given level.
+   * @param level The level the SAT solver backtracked to.
    */
   void notifyBacktrack(uint32_t nlevels);
```
Comment (ajreynol) [inline src/prop/theory_preregistrar.h:67]:
```suggestion
  void notifyBacktrack(uint32_t level);
```

File: src/prop/sat_solver.h [hunk 3]
```diff
@@ -160,6 +161,7 @@ class CDCLTSatSolver : public SatSolver
 
   /**
    * Return the current list of decisions made by the SAT solver.
+   * TODO: this should return a reference
```
Comment (ajreynol) [inline src/prop/sat_solver.h:164]: Link to a issue or remove.

File: src/prop/sat_solver.h [hunk 4]
```diff
@@ -171,6 +173,8 @@ class CDCLTSatSolver : public SatSolver
 
   virtual std::shared_ptr<ProofNode> getProof() = 0;
 
+  virtual SatProofManager* getProofManager() = 0;
```
Comment (ajreynol) [inline src/prop/sat_solver.h:176]: You can mark this as temporary.

**Review by ajreynol — CHANGES_REQUESTED** [review]
Just some minor comments.

**Review by ajreynol — APPROVED** [review]

**Comment by mpreiner** [comment]
results on non-incremental benchmarks with 300s timeout and 8G memory limit:
```
          total  solved     sat   unsat  timeout  memout  uniq   time_cpu
config
cadical  434212  374917  174158  200759    39501     385  4951 13209697.8
minisat  434212  372720  171946  200774    41716     335  2754 13989033.2
```

**Pull request merged**